// middlewear/esm5/lib/middlewear.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule, InjectionToken } from '@angular/core';
import { MiddlewareService } from './middlewear.service';
import { HttpClientModule } from '@angular/common/http';
/** @type {?} */
export var API_HOST = new InjectionToken('API_HOST');
var MiddlewearModule = /** @class */ (function () {
    function MiddlewearModule() {
    }
    /**
     * @param {?} host
     * @return {?}
     */
    MiddlewearModule.forRoot = /**
     * @param {?} host
     * @return {?}
     */
    function (host) {
        return {
            ngModule: MiddlewearModule,
            providers: [{
                    provide: API_HOST,
                    useValue: host
                }]
        };
    };
    MiddlewearModule.decorators = [
        { type: NgModule, args: [{
                    imports: [HttpClientModule],
                    providers: [MiddlewareService],
                },] }
    ];
    return MiddlewearModule;
}());
export { MiddlewearModule };
